File: sha3-256-arm64-macos/bench/bench.swift
import CryptoKit
import Foundation

private let digestBytes = Int(SHA3_256_DIGEST_BYTES)
private let mib = 1024.0 * 1024.0

private struct Run {
	let size: Int
	let iterations: Int
}

private let runs = [
	Run(size: 0, iterations: 200_000),
	Run(size: 64, iterations: 200_000),
	Run(size: 1_024, iterations: 50_000),
	Run(size: 16_384, iterations: 10_000),
	Run(size: 262_144, iterations: 1_000),
	Run(size: 1_048_576, iterations: 300),
	Run(size: 10_485_760, iterations: 30),
]

private let columns: [(header: String, width: Int)] = [
	("size(B)", 9),
	("iters", 8),
	("asm(ms)", 10),
	("asm h/s", 10),
	("asm MiB/s", 11),
	("CryptoKit(ms)", 14),
	("CryptoKit h/s", 13),
	("CryptoKit MiB/s", 16),
	("OpenSSL(ms)", 12),
	("OpenSSL h/s", 13),
	("OpenSSL MiB/s", 15),
	("asm/CryptoKit", 13),
	("asm/OpenSSL", 12),
]

private struct Metrics {
	let milliseconds: Double
	let hashesPerSecond: Double
	let mibPerSecond: Double
}

private func pad(_ value: String, to width: Int) -> String {
	if value.count >= width {
		return value
	}
	return String(repeating: " ", count: width - value.count) + value
}

private func row(_ cells: [String]) -> String {
	zip(cells, columns).map { pad($0.0, to: $0.1.width) }.joined(separator: "  ")
}

@inline(never)
private func time(iterations: Int, body: () -> Void) -> Double {
	for _ in 0..<min(1_000, iterations) {
		body()
	}
	let start = ContinuousClock.now
	for _ in 0..<iterations {
		body()
	}
	let duration = start.duration(to: ContinuousClock.now)
	let components = duration.components
	return Double(components.seconds) * 1_000.0
		+ Double(components.attoseconds) / 1.0e15
}

private func metrics(elapsedMilliseconds: Double, run: Run) -> Metrics {
	let seconds = max(elapsedMilliseconds / 1_000.0, Double.leastNonzeroMagnitude)
	return Metrics(
		milliseconds: elapsedMilliseconds,
		hashesPerSecond: Double(run.iterations) / seconds,
		mibPerSecond: (Double(run.size * run.iterations) / mib) / seconds
	)
}

private func format(_ metrics: Metrics) -> [String] {
	[
		String(format: "%.2f", metrics.milliseconds),
		String(format: "%.0f", metrics.hashesPerSecond),
		String(format: "%.2f", metrics.mibPerSecond),
	]
}

private final class Buffers {
	let context: UnsafeMutablePointer<sha3_256_ctx>
	let message: UnsafeMutablePointer<UInt8>
	let digest: UnsafeMutablePointer<UInt8>
	let messageSize: Int

	init(size: Int) {
		context = .allocate(capacity: 1)
		message = .allocate(capacity: max(size, 1))
		digest = .allocate(capacity: digestBytes)
		messageSize = size
		message.initialize(repeating: 0xa3, count: max(size, 1))
		digest.initialize(repeating: 0, count: digestBytes)
	}

	deinit {
		digest.deinitialize(count: digestBytes)
		digest.deallocate()
		message.deinitialize(count: max(messageSize, 1))
		message.deallocate()
		context.deallocate()
	}

	var rawMessage: UnsafeRawBufferPointer {
		UnsafeRawBufferPointer(start: message, count: messageSize)
	}

	func digestArray() -> [UInt8] {
		Array(UnsafeBufferPointer(start: digest, count: digestBytes))
	}
}

private final class OpenSSLHash {
	let algorithm: OpaquePointer
	let context: OpaquePointer

	init() {
		guard let algorithm = EVP_MD_fetch(nil, "SHA3-256", nil),
			let context = EVP_MD_CTX_new()
		else {
			fatalError("OpenSSL could not initialize SHA3-256")
		}
		self.algorithm = algorithm
		self.context = context
	}

	deinit {
		EVP_MD_CTX_free(context)
		EVP_MD_free(algorithm)
	}

	@inline(never)
	func hash(_ message: UnsafeRawBufferPointer, into output: UnsafeMutablePointer<UInt8>) {
		var outputLength: UInt32 = 0
		precondition(EVP_DigestInit_ex2(context, algorithm, nil) == 1)
		precondition(EVP_DigestUpdate(context, message.baseAddress, message.count) == 1)
		precondition(EVP_DigestFinal_ex(context, output, &outputLength) == 1)
		precondition(outputLength == digestBytes)
	}
}

@inline(never)
private func assemblyHash(_ buffers: Buffers) {
	sha3_256_init(buffers.context)
	sha3_256_update(buffers.context, buffers.rawMessage.baseAddress, buffers.messageSize)
	sha3_256_digest(buffers.context, buffers.digest)
}

@inline(never)
private func cryptoKitHash(_ buffers: Buffers) {
	var hash = SHA3_256()
	hash.update(bufferPointer: buffers.rawMessage)
	let digest = hash.finalize()
	digest.withUnsafeBytes { bytes in
		buffers.digest.update(from: bytes.bindMemory(to: UInt8.self).baseAddress!,
			count: digestBytes)
	}
}

@main
private struct Benchmark {
	static func main() {
		let openssl = OpenSSLHash()
		print(row(columns.map(\.header)))

		for run in runs {
			let buffers = Buffers(size: run.size)

			assemblyHash(buffers)
			let assemblyDigest = buffers.digestArray()
			cryptoKitHash(buffers)
			let cryptoKitDigest = buffers.digestArray()
			openssl.hash(buffers.rawMessage, into: buffers.digest)
			let opensslDigest = buffers.digestArray()
			guard assemblyDigest == cryptoKitDigest, assemblyDigest == opensslDigest else {
				fputs("digest mismatch at size \(run.size)\n", stderr)
				exit(EXIT_FAILURE)
			}

			let assembly = metrics(
				elapsedMilliseconds: time(iterations: run.iterations) {
					assemblyHash(buffers)
				},
				run: run
			)
			let cryptoKit = metrics(
				elapsedMilliseconds: time(iterations: run.iterations) {
					cryptoKitHash(buffers)
				},
				run: run
			)
			let opensslMetrics = metrics(
				elapsedMilliseconds: time(iterations: run.iterations) {
					openssl.hash(buffers.rawMessage, into: buffers.digest)
				},
				run: run
			)

			print(row(
				[String(run.size), String(run.iterations)]
					+ format(assembly)
					+ format(cryptoKit)
					+ format(opensslMetrics)
					+ [
						String(format: "%.2fx", cryptoKit.milliseconds / assembly.milliseconds),
						String(format: "%.2fx", opensslMetrics.milliseconds / assembly.milliseconds),
					]
			))
		}
	}
}

File: sha3-256-x86_64-linux/bench/bench.c
#define _POSIX_C_SOURCE 199309L

#include "../src/sha3-256.h"

#include <gcrypt.h>
#include <openssl/evp.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static const size_t digest_bytes = SHA3_256_DIGEST_BYTES;
static const double mib = 1024.0 * 1024.0;

struct run {
	size_t size;
	int iterations;
};

static const struct run runs[] = {
	{0, 200000},
	{64, 200000},
	{1024, 50000},
	{16384, 10000},
	{262144, 1000},
	{1048576, 300},
	{10485760, 30},
};

struct column {
	const char *header;
	int width;
};

static const struct column columns[] = {
	{"size(B)", 9},
	{"iters", 8},
	{"asm(ms)", 10},
	{"asm h/s", 10},
	{"asm MiB/s", 11},
	{"OpenSSL(ms)", 12},
	{"OpenSSL h/s", 13},
	{"OpenSSL MiB/s", 15},
	{"gcrypt(ms)", 11},
	{"gcrypt h/s", 12},
	{"gcrypt MiB/s", 14},
	{"asm/OpenSSL", 12},
	{"asm/gcrypt", 11},
};

enum { column_count = sizeof(columns) / sizeof(columns[0]) };

struct metrics {
	double milliseconds;
	double hashes_per_second;
	double mib_per_second;
};

struct timing {
	/* Time for the full iteration count at the fastest repetition's rate. */
	double milliseconds;
	/* (slowest - fastest) / fastest across the repetitions. */
	double spread;
};

struct buffers {
	sha3_256_ctx context;
	uint8_t *message;
	size_t size;
	uint8_t digest[SHA3_256_DIGEST_BYTES];
};

static double now_milliseconds(void) {
	struct timespec ts;
	clock_gettime(CLOCK_MONOTONIC, &ts);
	return (double)ts.tv_sec * 1000.0 + (double)ts.tv_nsec / 1.0e6;
}

/*
 * The warmup runs until the core has sustained load long enough to reach
 * its maximum frequency; a fixed iteration count is too short for cheap
 * bodies. Reporting the fastest of five equal repetitions discards
 * scheduling and frequency-ramp noise, which only ever slows a run down.
 */
static struct timing time_hash(int iterations, void (*body)(struct buffers *),
	struct buffers *buffers) {
	const int repetitions = 5;
	int per_repetition = iterations / repetitions;
	if (per_repetition < 1) {
		per_repetition = 1;
	}
	double warmup_start = now_milliseconds();
	while (now_milliseconds() - warmup_start < 200.0) {
		body(buffers);
	}
	double fastest = 0.0;
	double slowest = 0.0;
	for (int repetition = 0; repetition < repetitions; ++repetition) {
		double start = now_milliseconds();
		for (int iteration = 0; iteration < per_repetition; ++iteration) {
			body(buffers);
		}
		double elapsed = now_milliseconds() - start;
		if (repetition == 0 || elapsed < fastest) {
			fastest = elapsed;
		}
		if (elapsed > slowest) {
			slowest = elapsed;
		}
	}
	struct timing timing = {
		fastest * (double)iterations / (double)per_repetition,
		(slowest - fastest) / fastest,
	};
	return timing;
}

static struct metrics metrics_for(double milliseconds, const struct run *run) {
	double seconds = milliseconds / 1000.0;
	if (seconds <= 0.0) {
		seconds = 1.0e-300;
	}
	struct metrics metrics = {
		milliseconds,
		(double)run->iterations / seconds,
		((double)run->size * (double)run->iterations / mib) / seconds,
	};
	return metrics;
}

static EVP_MD *openssl_algorithm;
static EVP_MD_CTX *openssl_context;
static gcry_md_hd_t gcrypt_handle;

__attribute__((noinline))
static void assembly_hash(struct buffers *buffers) {
	sha3_256_init(&buffers->context);
	sha3_256_update(&buffers->context, buffers->message, buffers->size);
	sha3_256_digest(&buffers->context, buffers->digest);
}

__attribute__((noinline))
static void openssl_hash(struct buffers *buffers) {
	unsigned int output_length = 0;
	if (EVP_DigestInit_ex2(openssl_context, openssl_algorithm, NULL) != 1 ||
		EVP_DigestUpdate(openssl_context, buffers->message,
			buffers->size) != 1 ||
		EVP_DigestFinal_ex(openssl_context, buffers->digest,
			&output_length) != 1 ||
		output_length != digest_bytes) {
		fprintf(stderr, "OpenSSL SHA3-256 failed\n");
		exit(EXIT_FAILURE);
	}
}

__attribute__((noinline))
static void gcrypt_hash(struct buffers *buffers) {
	gcry_md_reset(gcrypt_handle);
	gcry_md_write(gcrypt_handle, buffers->message, buffers->size);
	memcpy(buffers->digest, gcry_md_read(gcrypt_handle, GCRY_MD_SHA3_256),
		digest_bytes);
}

static void pad(const char *value, int width) {
	printf("%*s", width, value);
}

static void print_row(const char *cells[column_count]) {
	for (int index = 0; index < column_count; ++index) {
		if (index != 0) {
			printf("  ");
		}
		pad(cells[index], columns[index].width);
	}
	printf("\n");
}

static void format_metrics(const struct metrics *metrics,
	char cells[3][32]) {
	snprintf(cells[0], 32, "%.2f", metrics->milliseconds);
	snprintf(cells[1], 32, "%.0f", metrics->hashes_per_second);
	snprintf(cells[2], 32, "%.2f", metrics->mib_per_second);
}

int main(void) {
	if (gcry_check_version(NULL) == NULL) {
		fprintf(stderr, "libgcrypt version check failed\n");
		return EXIT_FAILURE;
	}
	gcry_control(GCRYCTL_DISABLE_SECMEM, 0);
	gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
	if (gcry_md_open(&gcrypt_handle, GCRY_MD_SHA3_256, 0) != 0) {
		fprintf(stderr, "libgcrypt could not initialize SHA3-256\n");
		return EXIT_FAILURE;
	}
	openssl_algorithm = EVP_MD_fetch(NULL, "SHA3-256", NULL);
	openssl_context = EVP_MD_CTX_new();
	if (openssl_algorithm == NULL || openssl_context == NULL) {
		fprintf(stderr, "OpenSSL could not initialize SHA3-256\n");
		return EXIT_FAILURE;
	}

	double worst_spreads[3] = {0.0, 0.0, 0.0};
	const char *headers[column_count];
	for (int index = 0; index < column_count; ++index) {
		headers[index] = columns[index].header;
	}
	print_row(headers);

	for (size_t run_index = 0; run_index < sizeof(runs) / sizeof(runs[0]);
		++run_index) {
		const struct run *run = &runs[run_index];
		struct buffers buffers;
		buffers.size = run->size;
		buffers.message = malloc(run->size > 0 ? run->size : 1);
		if (buffers.message == NULL) {
			fprintf(stderr, "allocation failed at size %zu\n", run->size);
			return EXIT_FAILURE;
		}
		memset(buffers.message, 0xa3, run->size > 0 ? run->size : 1);

		uint8_t assembly_digest[SHA3_256_DIGEST_BYTES];
		uint8_t openssl_digest[SHA3_256_DIGEST_BYTES];
		assembly_hash(&buffers);
		memcpy(assembly_digest, buffers.digest, digest_bytes);
		openssl_hash(&buffers);
		memcpy(openssl_digest, buffers.digest, digest_bytes);
		gcrypt_hash(&buffers);
		if (memcmp(assembly_digest, openssl_digest, digest_bytes) != 0 ||
			memcmp(assembly_digest, buffers.digest, digest_bytes) != 0) {
			fprintf(stderr, "digest mismatch at size %zu\n", run->size);
			return EXIT_FAILURE;
		}

		struct timing timings[3];
		timings[0] = time_hash(run->iterations, assembly_hash, &buffers);
		timings[1] = time_hash(run->iterations, openssl_hash, &buffers);
		timings[2] = time_hash(run->iterations, gcrypt_hash, &buffers);
		for (int index = 0; index < 3; ++index) {
			if (timings[index].spread > worst_spreads[index]) {
				worst_spreads[index] = timings[index].spread;
			}
		}
		struct metrics assembly =
			metrics_for(timings[0].milliseconds, run);
		struct metrics openssl_metrics =
			metrics_for(timings[1].milliseconds, run);
		struct metrics gcrypt_metrics =
			metrics_for(timings[2].milliseconds, run);

		char size_cell[32];
		char iters_cell[32];
		char metric_cells[3][3][32];
		char ratio_cells[2][32];
		snprintf(size_cell, sizeof(size_cell), "%zu", run->size);
		snprintf(iters_cell, sizeof(iters_cell), "%d", run->iterations);
		format_metrics(&assembly, metric_cells[0]);
		format_metrics(&openssl_metrics, metric_cells[1]);
		format_metrics(&gcrypt_metrics, metric_cells[2]);
		snprintf(ratio_cells[0], 32, "%.2fx",
			openssl_metrics.milliseconds / assembly.milliseconds);
		snprintf(ratio_cells[1], 32, "%.2fx",
			gcrypt_metrics.milliseconds / assembly.milliseconds);

		const char *cells[column_count] = {
			size_cell, iters_cell,
			metric_cells[0][0], metric_cells[0][1], metric_cells[0][2],
			metric_cells[1][0], metric_cells[1][1], metric_cells[1][2],
			metric_cells[2][0], metric_cells[2][1], metric_cells[2][2],
			ratio_cells[0], ratio_cells[1],
		};
		print_row(cells);
		free(buffers.message);
	}

	printf("worst repetition spread: asm %.1f%%, OpenSSL %.1f%%, "
		"gcrypt %.1f%%\n",
		worst_spreads[0] * 100.0, worst_spreads[1] * 100.0,
		worst_spreads[2] * 100.0);

	gcry_md_close(gcrypt_handle);
	EVP_MD_CTX_free(openssl_context);
	EVP_MD_free(openssl_algorithm);
	return EXIT_SUCCESS;
}
